query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRestClientFactory.java

```diff
@@ -0,0 +1,43 @@
+package org.hypertrace.core.query.service.prometheus;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+class PrometheusRestClientFactory {
+
+  // Singleton instance
+  private static final PrometheusRestClientFactory INSTANCE = new PrometheusRestClientFactory();
```

Comment:
Aren't we already using guice?

Reply:
Tried using it, isn't looking very clean

Reply:
Hmm - IMO, the current form where we implement a singleton by hand, use statics, call a static method and ignore its return to modify state and then needlessly take in a handler name is significantly harder to understand and objectively more code. There's potential bugs in there too, because if I give a handler name that's already been registered, my connection string is ignored.


```diff
diff --git a/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandler.java b/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandler.java
index 9f76633..355bbf3 100644
--- a/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandler.java
+++ b/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandler.java
@@ -36,13 +36,14 @@ public class PrometheusBasedRequestHandler implements RequestHandler {
   private Optional<String> startTimeAttributeName;
   private PrometheusViewDefinition prometheusViewDefinition;
 
-  PrometheusBasedRequestHandler(String name, Config requestHandlerConfig) {
+  PrometheusBasedRequestHandler(
+      String name, Config requestHandlerConfig, PrometheusRestClient prometheusRestClient) {
     this.name = name;
     this.processConfig(requestHandlerConfig);
     this.queryRequestEligibilityValidator =
         new QueryRequestEligibilityValidator(prometheusViewDefinition);
     this.requestToPromqlConverter = new QueryRequestToPromqlConverter(prometheusViewDefinition);
-    this.prometheusRestClient = PrometheusRestClientFactory.get().getPrometheusClient(name);
+    this.prometheusRestClient = prometheusRestClient;
   }
 
   @Override
diff --git a/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRequestHandlerBuilder.java b/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRequestHandlerBuilder.java
index 4136015..acd0be7 100644
--- a/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRequestHandlerBuilder.java
+++ b/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRequestHandlerBuilder.java
@@ -10,10 +10,14 @@ import org.hypertrace.core.query.service.RequestHandlerClientConfigRegistry;
 public class PrometheusRequestHandlerBuilder implements RequestHandlerBuilder {
 
   private final RequestHandlerClientConfigRegistry clientConfigRegistry;
+  private final PrometheusRestClientFactory prometheusRestClientFactory;
 
   @Inject
-  PrometheusRequestHandlerBuilder(RequestHandlerClientConfigRegistry clientConfigRegistry) {
+  PrometheusRequestHandlerBuilder(
+      RequestHandlerClientConfigRegistry clientConfigRegistry,
+      PrometheusRestClientFactory prometheusRestClientFactory) {
     this.clientConfigRegistry = clientConfigRegistry;
+    this.prometheusRestClientFactory = prometheusRestClientFactory;
   }
 
   @Override
@@ -32,9 +36,9 @@ public class PrometheusRequestHandlerBuilder implements RequestHandlerBuilder {
                     new UnsupportedOperationException(
                         "Client config requested but not registered: " + config.getClientConfig()));
 
-    PrometheusRestClientFactory.createPrometheusClient(
-        config.getName(), clientConfig.getConnectionString());
-
-    return new PrometheusBasedRequestHandler(config.getName(), config.getRequestHandlerInfo());
+    return new PrometheusBasedRequestHandler(
+        config.getName(),
+        config.getRequestHandlerInfo(),
+        prometheusRestClientFactory.getPrometheusClient(clientConfig.getConnectionString()));
   }
 }
diff --git a/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRestClientFactory.java b/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRestClientFactory.java
index a1b7bad..e1c8d60 100644
--- a/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRestClientFactory.java
+++ b/query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusRestClientFactory.java
@@ -1,43 +1,15 @@
 package org.hypertrace.core.query.service.prometheus;
 
 import java.util.concurrent.ConcurrentHashMap;
+import javax.inject.Singleton;
 
+@Singleton
 class PrometheusRestClientFactory {
 
-  // Singleton instance
-  private static final PrometheusRestClientFactory INSTANCE = new PrometheusRestClientFactory();
-
   private final ConcurrentHashMap<String, PrometheusRestClient> clientMap =
       new ConcurrentHashMap<>();
 
-  private PrometheusRestClientFactory() {}
-
-  // Create a Prometheus Client.
-  public static PrometheusRestClient createPrometheusClient(
-      String requestHandlerName, String connectionString) {
-    if (!get().containsClient(requestHandlerName)) {
-      synchronized (get()) {
-        if (!get().containsClient(requestHandlerName)) {
-          get().addPrometheusClient(requestHandlerName, new PrometheusRestClient(connectionString));
-        }
-      }
-    }
-    return get().getPrometheusClient(requestHandlerName);
-  }
-
-  public static PrometheusRestClientFactory get() {
-    return INSTANCE;
-  }
-
-  private void addPrometheusClient(String requestHandlerName, PrometheusRestClient client) {
-    this.clientMap.put(requestHandlerName, client);
-  }
-
-  public boolean containsClient(String requestHandlerName) {
-    return this.clientMap.containsKey(requestHandlerName);
-  }
-
-  public PrometheusRestClient getPrometheusClient(String requestHandlerName) {
-    return this.clientMap.get(requestHandlerName);
+  public PrometheusRestClient getPrometheusClient(String connectionString) {
+    return clientMap.computeIfAbsent(connectionString, PrometheusRestClient::new);
   }
 }
diff --git a/query-service-impl/src/test/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandlerTest.java b/query-service-impl/src/test/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandlerTest.java
index 26a4607..b4d4c09 100644
--- a/query-service-impl/src/test/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandlerTest.java
+++ b/query-service-impl/src/test/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandlerTest.java
@@ -40,11 +40,12 @@ class PrometheusBasedRequestHandlerTest {
     mockWebServer.start(9099);
 
     Config config = PrometheusTestUtils.getDefaultPrometheusConfig();
-    PrometheusRestClientFactory.createPrometheusClient(
-        config.getString(CONFIG_PATH_NAME), config.getString(CONFIG_PATH_CLIENT_KEY));
+
     prometheusBasedRequestHandler =
         new PrometheusBasedRequestHandler(
-            config.getString(CONFIG_PATH_NAME), config.getConfig(CONFIG_PATH_REQUEST_HANDLER_INFO));
+            config.getString(CONFIG_PATH_NAME),
+            config.getConfig(CONFIG_PATH_REQUEST_HANDLER_INFO),
+            new PrometheusRestClient(config.getString(CONFIG_PATH_CLIENT_KEY)));
   }
 
   @AfterAll
```